File: NotesStudy/OC/Commons/HomePageMenu/HomePageMenuCell.m
//
//  HomePageMenuCell.m
//  NoteDome
//
//  Created by Lj on 2018/1/18.
//  Copyright © 2018年 Lj. All rights reserved.
//

#import "HomePageMenuCell.h"

@interface HomePageMenuCell ()

@end

@implementation HomePageMenuCell

- (void)setHighlighted:(BOOL)highlighted {
    [super setHighlighted:highlighted];
}


- (id)initWithFrame:(CGRect)frame {
    if (self = [super initWithFrame:frame]) {
        [self configDefaultValues];
        [self setUpUI];
        [self _viewLayout];
    }
    return self;
}


- (void)configDefaultValues {
    _iconImageSize = 50;
    _iconMarginTop = 10;
    _nameMarginTop = 5;
    _nameFont = [UIFont systemFontOfSize:15];
    _nameColor = [UIColor whiteColor];
}

#pragma mark -
- (void)setUpUI {
    self.iconImage = [[UIImageView alloc]init];
    self.iconImage.clipsToBounds = YES;
    self.iconImage.contentMode = UIViewContentModeScaleAspectFill;
    [self.contentView addSubview:self.iconImage];
    
    self.nameLabel = [[UILabel alloc]init];
    self.nameLabel.font = _nameFont;
    self.nameLabel.textColor = _nameColor;
    self.nameLabel.adjustsFontSizeToFitWidth = YES;
    self.nameLabel.textAlignment = NSTextAlignmentCenter;
    [self.contentView addSubview:self.nameLabel];
    
    self.numberLabel = [[UILabel alloc]init];
    self.numberLabel.textColor = [UIColor whiteColor];
    self.numberLabel.backgroundColor = [UIColor orangeColor];
    self.numberLabel.clipsToBounds = YES;
    self.numberLabel.layer.cornerRadius = 9;
    self.numberLabel.hidden = YES;
    self.numberLabel.font = [UIFont systemFontOfSize:12];
    self.numberLabel.textAlignment = NSTextAlignmentCenter;
    [self.contentView addSubview:self.numberLabel];
}

- (void)_viewLayout {
    [self.iconImage mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerX.equalTo(self.contentView);
        make.top.equalTo(self.contentView).offset(_iconMarginTop);
        make.width.height.offset(_iconImageSize);
    }];
    
    [self.nameLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerX.equalTo(self.contentView);
        make.left.equalTo(self.contentView).offset(5);
        make.top.equalTo(self.iconImage.mas_bottom).offset(_nameMarginTop);
    }];
    
    [self.numberLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.height.width.offset(18);
        make.left.equalTo(self.iconImage.mas_right).offset(- 9);
        make.top.equalTo(self.iconImage).offset(- 9);
    }];
}

- (void)setNumberStr:(NSString *)numberStr {
    _numberStr = numberStr;
    NSInteger number = [numberStr integerValue];
    if (number == 0) {
        self.numberLabel.hidden = YES;
    }else {
        self.numberLabel.hidden = NO;
        if (number > 99) {
            self.numberLabel.text = @"99+";
            [self.numberLabel mas_updateConstraints:^(MASConstraintMaker *make) {
                make.width.offset(27);
            }];
        }else if (number > 9) {
            self.numberLabel.text = numberStr;
            [self.numberLabel mas_updateConstraints:^(MASConstraintMaker *make) {
                make.width.offset(22);
            }];
        }else {
            self.numberLabel.text = numberStr;
            [self.numberLabel mas_updateConstraints:^(MASConstraintMaker *make) {
                make.width.offset(18);
            }];
        }
    }
}

- (void)setIconImageSize:(CGFloat)iconImageSize {
    _iconImageSize = iconImageSize;
    [self.iconImage mas_updateConstraints:^(MASConstraintMaker *make) {
        make.height.width.offset(iconImageSize);
    }];
}

//- (void)setNameColor:(UIColor *)nameColor {
//    _nameColor = nameColor;
//}

//- (void)setNameFont:(UIFont *)nameFont {
//    _nameFont = nameFont;
//}


@end

File: SwiftNoteStudy/SwiftNoteStudy/Commons/HomePageMenu/HomePageMenuCell.swift
//
//  HomePageMenuCell.swift
//  SwiftNoteStudy
//
//  Created by less on 2018/11/2.
//  Copyright © 2018 lj. All rights reserved.
//

import UIKit
import SnapKit

class HomePageMenuCell: UICollectionViewCell {
    
    public var number: Int = 0 {
        didSet {
            if number == 0 {
                numberLabel.isHidden = true
            }else {
                numberLabel.isHidden = false
                numberLabel.text = "\(number)"
                switch number {
                case 0..<10: numberWidth = 18
                case 10..<100: numberWidth = 22
                default: numberWidth = 27; numberLabel.text = "99+"
                }
            }
        }
    }
    
    public var cellStruct: MenuCellStruct = MenuCellStruct() {
        didSet {
            setNeedsDisplay()
        }
    }
    
    public var nameLabel: UILabel!
    public var iconImage: UIImageView!
    fileprivate var numberLabel: UILabel!
    fileprivate var numberWidth: CGFloat = 18
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        initView()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    fileprivate func initView() {
        iconImage = UIImageView()
        iconImage.backgroundColor = UIColor.red
        iconImage.contentMode = .scaleAspectFill
        iconImage.clipsToBounds = true
        contentView.addSubview(iconImage)
        
        nameLabel = UILabel()
        nameLabel.adjustsFontSizeToFitWidth = true
        nameLabel.textAlignment = .center
        contentView.addSubview(nameLabel)
        
        numberLabel = UILabel()
        numberLabel.textColor = UIColor.white
        numberLabel.backgroundColor = UIColor.blue
        numberLabel.layer.cornerRadius = 9
        numberLabel.font = UIFont.systemFont(ofSize: 12)
        numberLabel.clipsToBounds = true
        numberLabel.isHidden = true
        numberLabel.textAlignment = .center
        contentView.addSubview(numberLabel)
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        iconImage.frame = CGRect(x: (bounds.width - cellStruct.iconImageHeight)/2, y: cellStruct.iconImageTop, width: cellStruct.iconImageHeight, height: cellStruct.iconImageHeight)
        
        let titleHeight = (bounds.height - cellStruct.iconImageHeight - cellStruct.iconImageTop - cellStruct.titleTopHeight)
        nameLabel.frame = CGRect(x: 5, y: iconImage.frame.maxY + cellStruct.titleTopHeight, width: bounds.width - 10, height: titleHeight)
        nameLabel.font = cellStruct.titleFont
        nameLabel.textColor = cellStruct.titleColor
        
        let numberTop = cellStruct.iconImageTop > 9 ? iconImage.frame.minY - 9 : 0
        numberLabel.frame = CGRect(x: iconImage.frame.maxX - 9, y: numberTop, width: numberWidth, height: 18)
        
    }
    
}
